project4/proto/pub_sub_messages.proto

```diff
@@ -0,0 +1,140 @@
+syntax = "proto3";
+
+package pub_sub_messages;
+
+/// look up result to be rippled through the name servers
+message LookUpResult {
+  /// the key of the value to retrieve
+  uint32 key = 1;
+  /// the value associated with the key
+  string value = 2;
+  /// the id's of each contacted server
+  repeated uint32 server_ids = 3;
+  /// the id of the server storing the value
+  uint32 id = 4;
+}
+
+/// insert result to be rippled through the name servers
+message InsertResult {
+  /// the key of the value to insert
+  uint32 key = 1;
+  /// the value to insert
+  string value = 2;
+  /// the id of the name server the key/val pair was inserted into
+  uint32 id = 3;
+  /// the id's of each contacted server
+  repeated uint32 server_ids = 4;
+}
+
+/// delete result to be rippled through the name servers
+message DeleteResult {
+  /// the key of the key/val pair to delete
+  uint32 key = 1;
+  /// true if the key was successfully deleted, false if key was not found
+  bool delete_success = 2;
+  /// the id's of each contacted server
+  repeated uint32 server_ids = 3;
+}
+
+/// entrance request sent by the entering name server to the bootstrap name server.
+message EntranceRequest {
+  /// the id of the entering name server
+  uint32 id = 1;
+}
+
+/// entrance information to be rippled through the name servers.
+message EntranceInformation {
+  /// the id of the entering name server
+  uint32 id = 1;
+  /// the id of the successor of the entering name server
+  uint32 successor_id = 2;
```

Comment:
(minor) Organizationally, it might be a little clearer if you encapsulated `id`, `ip`, and `port` in a sub-message.